### ISSUE TITLE
job type base class 

### ISSUE THREAD

**Issue by anyaelena** [NREL/alfalfa#259]
setup (init + handers), run, teardown/cleanup (see boptest run def cleanup).

**Comment by kbenne** [comment]
I really like where this is going. I do have a few comments, which I will try to flesh out later today or tomorrow.

**Comment by kbenne** [comment]
> @anyaelena @nllong @kbenne Thoughts?
> 
> ## Job (extends AlfalfaConnectionsBase?)
Perhaps Job can simply become the root class and move anything we need from AlfalfaConnectionsBase into it? That name has always been a bit awkward to me.
> ### `__init__(args)`
> Takes care of setup actions which have no side-effects. Such as parsing args into variables
> 
> ### `setup()`
> Setup actions which do have side effects
> 
> * enumerate message callbacks (These will not be enabled until running so they could also be in the **init**)
> * create job folder structure
> * other job specific setup actions
> 
I think that future `Job` authors may have a hard time making the distinction between what should be done in `__init__` versus `setup`. Maybe authors of derived `Job` types should be encouraged to implement only `setup`. 
> ### `run()`
> Runs job (given the presence of callbacks maybe this should be start? as most of the job actions could be triggered through message callbacks)
I like the name `start` because it pairs well with `stop`. I imagine the base implementation would enter a loop that processes messages until a stop flag is detected. Is that what you have in mind? Some derived job types, such as the BOPTEST run job, might not need to reimplement this at all. Other job types (e.g "non-interactive" job types like `AddSite`) might reimplement and not call the base implementation. 

Would this call `cleanup` right before exit? Maybe the dispatcher would take care of that.
> 
> ### `cleanup()`
> Any cleanup action required for the job
> 
> * Delete handlers
> * cleanup files
> * package results?
> 
A base implementation that tars the working directory and pushes to storage would be nice. That seems like a universally useful thing and would provide consistency if done at the base level.
> ### `stop()`
> A couple approaches:
> 
> 1. Set a flag that the job should not continue (how boptest currently does it) and disable message callbacks. This requires that the job cares about the flag.
> 2. If jobs are run in threads we can just kill the thread.
> 
I vote for option 1. Perhaps `Jobs` run in threads at some point, but I feel like that is a thing for the caller to worry about.
> ### `status()`
> Base function that reads a status variable and reports it. Implementers only need to change the variable value in certain places to have the status accurately reported. Status variable could be String, Enum or a combination of the two
> 
Is there a common set of enums? I'm a bit skeptical of this feature.
> ## Docs
> Jobs could be "self documenting" in a way similar to openstudio measures. We could require a name and description function. Also, we could mandate handlers be declared in a way that allows us to get a list of messages and descriptions the job supports.
> 
I like this a lot! I prefer shorter methods like simply `name` and `description` though. I know it pertains to job, because I'm asking a job instance.
> ### `job_name()`
> job's name
> 
> ### `job_description()`
> job's description
> 
> ### `job_messages()`
> message callbacks are declared here along with mandatory description strings. That way we can scrape it for documentation, it can also be called by the base class to setup message callbacks.

**Comment by TShapinsky** [comment]
I've gone ahead and mocked up what I see as a possible job setup [here](https://github.com/TShapinsky/alfalfa_jobs_mockup)

If you just want to see the base class it is [here](https://github.com/TShapinsky/alfalfa_jobs_mockup/blob/main/alfalfa_jobs/job.py)

**Comment by kbenne** [comment]
@TShapinsky I looked through the mocked up class. I think it is beautiful. I do have some comments, but I need to prioritize other work before I can really dig in and provide substantive feedback. Basically I think this is the way, and any version of this I am 100% on board with.

Highlights:

* I like the use of a decorator to mark message handlers. (although there might be some tradeoffs that *might* cause pause)
* I like how you centralized the status enums. As long as we don't leave it up to Job type authors I think it will work really well and be consistent. 
* Is SimpleQueue a mock interface for redis?
* I'm not really sold on a threading feature. I guess the dispatcher would `start` and then peace out? Would it try to keep track of how many it started? Maybe prototyping how the dispatcher would interact with a threaded start method would illustrate some of the complexities and help inform the design.

**Comment by TShapinsky** [comment]
> @TShapinsky I looked through the mocked up class. I think it is beautiful. I do have some comments, but I need to prioritize other work before I can really dig in and provide substantive feedback. Basically I think this is the way, and any version of this I am 100% on board with.
Thanks Kyle 😃 
> 
> Highlights:
> 
>     * I like the use of a decorator to mark message handlers. (although there might be some tradeoffs that _might_ cause pause)
I think with a couple tweaks we should be able to address most use cases (i.e. adding params to callbacks, etc.)
> 
>     * I like how you centralized the status enums. As long as we don't leave it up to Job type authors I think it will work really well and be consistent.
> 
>     * Is SimpleQueue a mock interface for redis?
Yes, it's not very smart. I mostly just needed a way to test that callbacks were getting through and a queue provides a threadsafe way to send those test messages through. In an actual system this would not be needed.
> 
>     * I'm not really sold on a threading feature. I guess the dispatcher would `start` and then peace out? Would it try to keep track of how many it started? Maybe prototyping how the dispatcher would interact with a threaded start method would illustrate some of the complexities and help inform the design.
Yeah, I'm not sure threading is right for us. Right now there's an environment variable you can set if you want threaded jobs so it's very easily toggle-able. The main reason I'm using threads right now is that the whole mockup is single process, so without job threads I wouldn't be able to send messages or query status of the job, as it would just be tied up in the `_message_loop`.

The big advantage of a threaded job, in my opinion, is that is leaves the dispatcher the ability to 1. query job status and 2. deal with misbehaving or stuck jobs. The dispatcher would maintain a list of running jobs which could be queried by `job_id` and once a job finishes the thread can be joined and the job popped.

**Comment by TShapinsky** [comment]
Just some questions on my mind, putting them here to have a record.
* is it important to be able to setup a job (download files, make folders, etc. ) as a separate step from running a job? or is it fine to have them executed in order?
* if multiple jobs are run on the same site should the site be duplicated such that each job is acting on an untouched simulation directory? (does not apply to boptest)

**Comment by anyaelena** [comment]
Ok, I've done my first pass of catching up and i'm excited about the progress.  Thanks @TShapinsky!  I haven't fully absorbed this yet, to be honest, but one thing I want to think through is how cleanup/recovery and state tracking is managed for jobs that error out.  

Related to:

> * is it important to be able to setup a job (download files, make folders, etc. ) as a separate step from running a job? or is it fine to have them executed in order?
It's important that setup is separate, particularly because we want to move to multiple runs for a single uploaded file (#233)

> * if multiple jobs are run on the same site should the site be duplicated such that each job is acting on an untouched simulation directory? (does not apply to boptest)
I think the uploaded dir remains pristine, and the worker downloads the pristine copy each run.  We should work upload of the results file into Tim's deployment/logging setup.

**Comment by nllong** [comment]
I'm back from PTO and catching up. I like where this is going and I'm sure we will discuss more on tomorrow's call. 

Regarding `stop()`, I agree that it shouldn't kill the threads at the moment. If we do go down the threaded runners, then we will probably need to write plenty of tests to make sure threads are cleaned up properly.

Oh, and I love the use of classes as the jobs and the find_class method (https://github.com/TShapinsky/alfalfa_jobs_mockup/blob/main/alfalfa_jobs/dispatcher.py#L44-L50).

**Comment by anyaelena** [comment]
@nllong @kbenne @TShapinsky My notes from today.  

Naming:
- *model*: uploaded file
- *run*: a single run of the model
- *job*: a unit of work for worker
- not discussed: *worker*.  Above I meant an alfalfa-worker pod.

Eventually, we will build in ability to have many *run*s per *model*.  For job refactor, we will plan to keep 1-1.

Discussed:
- Are jobs always 1-1 with runs? conclude: no
- should simulation setup job (metadata generation) be separate from run job?  
    -  Considerations: 
         - we might want different setup someday (haystack, brick, ...).  
         -  if setup and run are separate jobs, they may be pulled by different workers.  setup must download model from S3, modify by running setup, and re-upload.  Then run job can download.  This is how it works now (minus the model/run distinction)
         - Kyle: clarify whether we want the same metadata (including point ids) for each model-run, or we want updated ids each run
    - Conclusion: leave setup and run as separate jobs for now and do re-upload to s3 after job runs. Revisit when we introduce many runs / model.

**Comment by kbenne** [comment]
Yeah I honestly don't know what is the right answer, but I think you are smart to be thinking about it. I'd wouldn't complain either way this goes. (Although experience with end users might cause reversal in either direction) 

I can say that at one point there was an issue in Alfalfa because we originally used the OS handles as the unique point references. The consequence was that if you resubmitted the same osm, you would get a duplicate site and a collision. We ended up abandoning that strategy and generating new uids every time the osm was uploaded / tagged.
